### DISCUSSION TITLE
sBTC Design - Peg-In UTXO Consolidation to a Single UTXO

### DISCUSSION THREAD

**Discussion by AshtonStephens** [stacks-archive/sbtc-developer-release#343] [Ideas]
This is a discussion that contains the design document for how the transfer handoff will be implemented for SIP-025, especially regarding to edge cases collating utxos.

**Comment by AshtonStephens** [comment]
Currently under investigation by @soju-drinker.

**Comment by AshtonStephens** [comment]
This is now a research discussion post around the details of the wallet peg-in requests such that after any bitcoin block there is a single UTXO to the peg wallet.

The decision to have a single UTXO to the peg-in wallet is a simplification of part of the SIP-025 protocol proposal that results in higher peg-in fees but a more robust system with clearer transaction fee ownership.

After discussions on 11/01/2023 we've decided that the simplest solution to paying for the consolidation of wallet peg-in UTXOs is to have every peg-in request pay for both the transaction and the inevitable fee of collating the peg-in transactions into a single transaction, wherein that additional fee is used instantly to pay for transaction consolidation. 

This means that the peg-in requests will cost approximately twice the transaction fee that the equivalent transfer would cost on the bitcoin blockchain as it will need to pay for roughly twice the data to be stored on the bitcoin blockchain.

**Comment by jcnelson** [comment]
The benefits of requiring a consolidation on every deposit are as follows:

* **Small wallet hand-off transactions.**  The size of a wallet hand-off transaction would be smaller than a miner's block-commit, and by extension, its fee can be smaller.  This is good, because if block-commits can get mined, then so can these.  Moreover, the Stackers could estimate the fee for this transaction by looking at ongoing block-commits.

* **Stacks nodes, not Stackers, can handle sBTC mints.** Because _every_ deposit must be consolidated, we can shift the responsibility for minting sBTC from the Stackers (via a specially-crafted contract-call transaction to `.sbtc`) to the Stacks node consensus rules.  To do this, we'd require a specially-crafted `OP_RETURN` output to be present in a consolidation transaction, which identifies it as such.  Then, Stacks nodes extract these transactions from Bitcoin blocks, and when they decode and parse them, they find the `OP_DROP` payload from the user's spent deposit tapscript.  From there, the node consensus rules would simply decode the `OP_DROP` and the tapscript into Clarity arguments, and directly call the mint function in `.sbtc`.  See [this diagram](https://github.com/stacks-network/sbtc/discussions/345#discussioncomment-7452311).  This is simpler because we already have the means of doing these things in the Stacks node; we don't have to go create bespoke code paths in `.sbtc` and the signer to craft and send `.sbtc` contract-calls.

* **The "right people" pay for wallet maintenance.**  By requiring consolidation on deposit, we address a major concern in the SIP regarding how to ensure that the wallet hand-off can be feasibly funded in an equitable manner.  By requiring users to pay for consolidating their UTXO immediately, we ensure that the cost to hand-off the wallet is (1) as small as can be and (2) independent of the number of users who have deposited or withdrawn.  In other words, immediate consolidation is more equitable -- users pay for the service only as they interact with it, and don't need to worry about contributing payment for the wallet hand-off (i.e. the expected gains from capturing the burnt BTC in the prepare phase for the purposes of financing wallet hand-off transactions ought to _vastly exceed_ the cost of the transaction fee).

**Comment by mijoco-btc** [comment]
This is awesome Jude ! 

If I understand correctly the signers need to sign each deposit (plus the current UTXO in with their peg wallet) and spend this back to their peg wallet with the new op_return ?

Two questions:

1) there's no way to guarantee that there will only be a single utxo in the peg wallet e.g. in case of a dust attack or even an `accidental` deposit directly to it above the dust limit. The design will need to cater for this?

2) Can the work/cost of the signers be reduced by batching (similar to discussion about withdrawals). If there were 100 deposits in a block could they spend all of them in a single transaction? The stacks node then sees 100 spent deposit tapscript payloads and initiates a mint for each. I'm assuming 1 tx spending 100 inputs is cheaper than 100 transactions but maybe the costs are similar - but this would also mean they wouldn't need to wait to avoid double spending the wallet utxo ?

**Comment by mijoco-btc** [comment]
@friedger 's suggestion to repurpose miner PoX btc burns to the peg wallet in order to offset fees might also be a consideration here. in that it will introduce new utxos which are not redeeming deposits.

**Comment by AshtonStephens** [comment]
@radicleart, @soju-drinker, @friedger  what additional information do we need to ascertain here to scope out this work? Do we need some kind of flow diagram?
